R/LearnerSurvCTree.R

```diff
@@ -0,0 +1,115 @@
+          ParamLgl$new("stump", default = FALSE, tags = "train"),
+          ParamLgl$new("lookahead", default = FALSE, tags = "train"),
+          ParamLgl$new("MIA", default = FALSE, tags = "train"),
+          ParamInt$new("nresample", lower = 1L, default = 9999L,
+            tags = "train"),
+          ParamDbl$new("tol", lower = 0, tags = "train"),
+          ParamInt$new("maxsurrogate", lower = 0L, default = 0L,
+            tags = "train"),
+          ParamLgl$new("numsurrogate", default = FALSE, tags = "train"),
+          ParamInt$new("mtry", lower = 0L, special_vals = list(Inf),
+            default = Inf, tags = "train"),
+          ParamInt$new("maxdepth", lower = 0L, special_vals = list(Inf),
+            default = Inf, tags = "train"),
+          ParamLgl$new("multiway", default = FALSE, tags = "train"),
+          ParamInt$new("splittry", lower = 0L, default = 2L, tags = "train"),
+          ParamLgl$new("intersplit", default = FALSE, tags = "train"),
+          ParamLgl$new("majority", default = FALSE, tags = "train"),
+          ParamLgl$new("caseweights", default = FALSE, tags = "train"),
+          ParamUty$new("applyfun", tags = "train"),
+          ParamInt$new("cores", special_vals = list(NULL), default = NULL,
+            tags = "train"),
+          ParamLgl$new("saveinfo", default = TRUE, tags = "train"),
+          ParamLgl$new("update", default = FALSE, tags = "train"),
+          ParamFct$new("splitflavour", default = "ctree",
+            levels = c("ctree", "exhaustive"), tags = c("train", "control")),
+          ParamUty$new("offset", tags = "train"),
+          ParamUty$new("cluster", tags = "train"),
+          ParamUty$new("scores", tags = "train"),
+          ParamLgl$new("doFit", default = TRUE, tags = "train"),
+          ParamUty$new("pargs", default = mvtnorm::GenzBretz, tags = "train")
+        )
+      )
+      ps$add_dep("nresample", "testtype", CondEqual$new("MonteCarlo"))
+
+      super$initialize(
+        id            = "surv.ctree",
+        packages      = c("partykit", "coin", "distr6"),
+        feature_types = c("integer", "numeric", "factor", "ordered"),
+        predict_types = c("distr", "crank"),
+        param_set     = ps,
+        properties    = "weights"
+      )
+    }
+  ),
+
+  private = list(
+    .train = function(task) {
+      pars = self$param_set$get_values(tags = "train")
+
+      if ("weights" %in% task$properties) {
+        pars$weights = task$weights$weight
+      }
+
+      mod <- mlr3misc::invoke(partykit::ctree, formula = task$formula(),
+        data = task$data(), .args = pars)
+    },
+
+    .predict = function(task) {
+      newdata = task$data(cols = task$feature_names)
+
+      p = mlr3misc::invoke(predict, self$model, type = "prob", newdata = newdata)
+
+      # Define WeightedDiscrete distr6 distribution from the survival function
+      x = lapply(p, function(z) data.frame(x = z$time, cdf = 1 - z$surv))
```

Comment:
would be interesting to microbenchmark my method of first constructing an empty list then using `for` versus directly using `lapply`. I'll check this

Reply:
Ran a microbenchmark on this, the `for` method is actually quicker. Comparing both on 1000 samples, `for` gives a mean of 20ms and `lapply` is 240ms. Would you mind changing to the `for` method then with the initialized list? See coxph for an example of this.

Reply:
the problem is that the predict function returns list of survfit objects, each with different dimension depending on covariate vector. The number of time points in each element is different and depends on the unique event times of observations that during training, ended up in that final node. Therefore I cannot initialise list with fixed "time" for all list elements.

Reply:
strange, I'd assumed they would do the usual thing of having all time-points in the training set, I wonder if this will have implications for distribution evaluation

Reply:
>I wonder if this will have implications for distribution evaluation

Can´t see why, as long as one uses a clean implementation and not specific numerical hacks that assume otherwise.

Reply:
So what is the prefered resolution here?

Reply:
keep what you did Adi, thanks